### DISCUSSION TITLE
Can the ESP32-S3 run programs on a W25Q02 (2G-BIT) or W25Q01 (1G-BIT)? Firmware.bin drag-and-drop?

### DISCUSSION THREAD

**Discussion by rtek1000** [espressif/arduino-esp32#9387] [General]
What is the maximum internal program SPI flash memory that the ESP32-S3 can handle?

I have seen more and more models putting 128M-BIT on boards for mass production (perhaps to try to gain market share in the face of so many options, RP2040, STM32, F1C100s, RV1103 etc.), sometimes a 256M-BIT is needed, but I can already imagine an ESP32 with a 2G-BIT flash ([W25Q02](https://www.mouser.com/ProductDetail/Winbond/W25Q02JVTBIM?qs=TuK3vfAjtkVTVCDKxnTswg%3D%3D) - 2 Giga bits).

Maybe the question is about the address, where the W25Q02 needs 32-BIT and the W25Q128 only 24-BIT.

I don't know how long it would take to transfer 256MB of data on the ESP32. Would it take >39 minutes to send the ["data"](https://github.com/me-no-dev/arduino-esp32fs-plugin) folder at 921600 bauds?

I saw that the ESP32-S3 accepts drag-and-drop via USB Host MSC (and it is much faster than the serial port), to update the firmware (In the Arduino IDE it is like "USB Firmware MSC On Boot"). DFU needs a tool software [dfu-util](https://docs.espressif.com/projects/esp-idf/en/stable/esp32s2/api-guides/dfu.html#api-guide-dfu-flash), and it seems not to be as interesting as MSC, which doesn't need anything other than a computer or other host that can manipulate data on an MSC device.

How to update the area where the "data" folder is, using drag-and-drop?

![ESP32_drag-and-drop](https://github.com/espressif/arduino-esp32/assets/8660811/374b5739-035e-46ae-a25c-a3b10a5bd14f)

![Data_upload](https://github.com/espressif/arduino-esp32/assets/8660811/1b0f5428-04eb-4e4d-a10c-6a7fc57f9061)

![Data_upload2](https://github.com/espressif/arduino-esp32/assets/8660811/9a183cee-371c-4992-93fd-3d6c33eef361)

**Comment by Jason2866** [comment]
For flash 32 bit address support see https://docs.espressif.com/projects/esp-idf/en/release-v5.1/esp32s3/api-reference/peripherals/spi_flash/spi_flash_optional_feature.html

**Comment by rtek1000** [comment]
> For flash 32 bit address support see https://docs.espressif.com/projects/esp-idf/en/release-v5.1/esp32s3/api-reference/peripherals/spi_flash/spi_flash_optional_feature.html


From [[Optional features for flash](https://docs.espressif.com/projects/esp-idf/en/release-v5.1/esp32s3/api-reference/peripherals/spi_flash/spi_flash_optional_feature.html)]:
> 32-bit Address Flash Chips
> 
> This feature is supported on all Espressif chips (with various restrictions to application).
> 
> Most NOR flash chips used by Espressif chips use 24-bits address, which can cover 16 MBytes memory. However, for larger memory (usually equal to or larger than 16 MBytes), flash uses a 32-bits address to address larger memory. Regretfully, 32-bits address chips have vendor-specific commands, so we need to support the chips one by one.
> 
> List of Flash chips that support this feature:
> 
>     W25Q256
> 
>     GD25Q256
> 
> 
> 
> Important
> 
> **Over 16 MBytes space on flash mentioned above can be only used for data saving**, like file system. If your data/instructions over 16 MBytes spaces need to be mapped to MMU (so as to be accessed by the CPU), please upgrade to ESP-IDF v5.2 and read the latest docs.

- Over 16 MBytes space on flash mentioned above can be only used for data saving
